### ISSUE TITLE
Long text in inverted Flatlist causes huge performance drop on Android

### ISSUE THREAD

**Issue by divonelnc** [facebook/react-native#30034]
## Description
I recently upgraded to react 0.63.2 thanks to the latest Expo SDK 39. I immediately noticed that my most important FlatLists took a huge performance hit on **Android** (iOs seem unaffected).

After investigating, I found out that it happens when using `inverted` on a FlatList that contains items with a lot of text (see snack).

The same Flatlist, with the same data, is perfectly smooth when not inverted.

I have yet to test it in production, as the latest Expo SDK is causing trouble that stops me from building the app.

## React Native version:
react-native: https://github.com/expo/react-native/archive/sdk-39.0.0.tar.gz => 0.63.2

## Steps To Reproduce
Provide a detailed list of steps that reproduce the issue.

1. Create a Flatlist that renders items that contain a lot of text
2. Set the Flatlist as inverted

## Expected Results
The Flatlist should be as smooth when inverted as when normal.

## Snack, code example, screenshot, or link to a repository:
https://snack.expo.io/@divone/4c8d68

**Comment by divonelnc** [comment]
Just checked it on production and the performance hit is still there, making the app feel laggy.

**Comment by divonelnc** [comment]
This issue appears as soon as you invert a ScrollView actually, as you can see on this snack: https://snack.expo.io/@divone/c0c1f3

Again, only visible on a physical Android device.

**Comment by divonelnc** [comment]
It seems that the issue only happens on some phones. For me, it happens on a Google Pixel 4 with Android 11.

**Comment by SuryaL** [comment]
I see this issue on Pixel4 XL running Android 11, which was working fine on React Native 0.61.5 with Android 10. I notice that when the keyboard is open it does not seem to lag.

**Comment by safaiyeh** [comment]
Hey @divonelnc thanks for the investigation! I noticed the same issue when running your example.

**Comment by cookiespam** [comment]
Facing this issue on a Samsung Galaxy Note10 running Android 11 as well with Expo SDK 39. Didn't happen on Android 10. 

Discovered that changing the `verticallyInverted` style prop from `transform: [{scaleY: -1}]` to `scaleY: -1` in `VirtualizedList` fixes it.

**Comment by chr-sk** [comment]
Experienced this issue as well. I upgraded my Samsung Galaxy S10 from Android 10 to Android 11 and suddenly experienced a noticable lag on the `inverted` FlatList. @cookiespam 's workaround fixed the problem, however.

**Comment by TwanLuttik** [comment]
Im experience the same issue on android with a Samsung Galaxy s10 Android 11, and i noticed after updated the same issue.

**Comment by vinceplusplus** [comment]
@cookiespam's workaround works. for a less intrusive fix, we could "invert" twice ourselves, once in the container and once in every cell
```js
return (
  <FlatList
    // ...
    style={{scaleY: -1}}
    renderItem={({item}) => {
      return (
        <View style={{scaleY: -1}}>
          {/* cell content */}
        </View
      );
    }}
  >
);
```
for ios, could just use the official `inverted` prop. however, that `scaleY` outside of `transform` seems undocumented. @cookiespam how did you come up with it? great thanks though

**Comment by cookiespam** [comment]
@vinceplusplus

What I did eventually was just simply 
 ```js
   <FlatList
     // ...
     style= {
       Platform.OS === 'android'
        ? { scaleY: -1 } : { transform: [{ scaleY: -1 }] }
     }
   />
```
same goes for the items inside the flatlist as appropriate. 

But I think we can also do something like

 ```js
   <FlatList
     // ...
     inverted={ Platform.OS === 'ios'}
     style= {
       Platform.OS === 'android'
        && { scaleY: -1 }
     }
   />
```

**Comment by cookiespam** [comment]
@vinceplusplus

I was just trying to invert the `FlatList` on my own without using the `inverted` prop and somehow accidentally used `scaleY `(was googling how to invert something using css) which had been deprecated.

**Comment by vinceplusplus** [comment]
@cookiespam then that was a good accident 👍

**Comment by pqkluan** [comment]
I also encounter this issue with my chat screen. It's really hard to find to root cause.
- Only happen on Android S10e device.
- With inverted FlatList.
- FlatList Item contains a Text with any given length.

It's only take 10 items in the list for the performance to be hit really bad.

I'm using @cookiespam bandaid + `patch-package` to resolve the issue for now now.

Here is the the content of the patch:

`react-native+0.63.3.patch`
```diff
diff --git a/node_modules/react-native/Libraries/Lists/VirtualizedList.js b/node_modules/react-native/Libraries/Lists/VirtualizedList.js
index 9ec105f..a8f1d8c 100644
--- a/node_modules/react-native/Libraries/Lists/VirtualizedList.js
+++ b/node_modules/react-native/Libraries/Lists/VirtualizedList.js
@@ -11,6 +11,7 @@
 'use strict';
 
 const Batchinator = require('../Interaction/Batchinator');
+const Platform = require('../Utilities/Platform');
 const FillRateHelper = require('./FillRateHelper');
 const PropTypes = require('prop-types');
 const React = require('react');
@@ -2185,9 +2186,14 @@ function describeNestedLists(childList: {
 }
 
 const styles = StyleSheet.create({
-  verticallyInverted: {
-    transform: [{scaleY: -1}],
-  },
+  verticallyInverted:
+    Platform.OS === 'android'
+      ? {
+          scaleY: -1,
+        }
+      : {
+          transform: [{scaleY: -1}]
+        },
   horizontallyInverted: {
     transform: [{scaleX: -1}],
   },

```

**Comment by OlivierFreyssinet** [comment]
Wow wow wow thank you guys this fix is a lifesaver !!! 🙏

I was actually trying to replace my FlatList by a ScrollView to see where the issue was coming from and noticed then that without inverting the FlatList the performance was fine.

Now I can keep my FlatList inverted, it works exactly same as before without changing any other props and I'm going from horrible 40-50fps (since updating to Android 11) to butter smooth 90fps 😍

the solution proposed by @vinceplusplus is what I will be using:
> ```js
> return (
>   <FlatList
>     // ...
>     style={{scaleY: -1}}
>     renderItem={({item}) => {
>       return (
>         <View style={{scaleY: -1}}>
>           {/* cell content */}
>         </View
>       );
>     }}
>   >
> );
> ```

Is someone considered making a pull request with @cookiespam and @pqkluan's patch? I could make one in a couple of days when I get some free time but I also don't want to steal your moment of glory haha

**Comment by Jovan1998** [comment]
> Facing this issue on a Samsung Galaxy Note10 running Android 11 as well with Expo SDK 39. Didn't happen on Android 10.
> 
> Discovered that changing the `verticallyInverted` style prop from `transform: [{scaleY: -1}]` to `scaleY: -1` in `VirtualizedList` fixes it for Android, however it does not invert on iOS.

@cookiespam You are legend, worked for me too.

**Comment by enahum** [comment]
using the `scaleY: -1` instead of `transform: [{scaleY: -1}]` indeed seems to fix the performance issue, **BUT** (as usual there is a but) the refresh control becomes an issue.

It renders on the top of the screen as you attempt to scroll down and in order to actually scroll you'll need to scroll up then down to "free" the pan responder somehow.

**Comment by JamesMahy** [comment]
> @vinceplusplus
> 
> I was just trying to invert the `FlatList` on my own without using the `inverted` prop and somehow accidentally used `scaleY `(was googling how to invert something using css) which had been deprecated.

I think I love you, thank you!

**Comment by KochMario** [comment]
I'd just wanted to drop in and say thank you as well 🙏🏼💙

I was severely confused when I upgraded to Android 11 (on a Redmi Note 8 Pro) and the performance on the chat screen in my app was suddenly horrible. Scrolling for a couple of seconds resulted in 1000+ dropped frames. It was so laggy and jerky that one got almost dizzy by watching it.

Based on your suggestions I've implemented a simple check (with react-native-device-info) so the fix is only applied for android & version >= 11.

```javascript
const useAndroidInvertedFix = useMemo(
  () => Platform.OS === 'android' && Number(getSystemVersion()) >= 11,
  [],
);
```

I'm extremely grateful for this workaround, however the underlying issue should be officially addressed by react-native itself, right?
Does anyone know the procedure for doing that or is it enough that this issue gets enough attention? 🤔

**Comment by safaiyeh** [comment]
@KochMario the best way to get this fix in is by submitting a PR. Unfortunately the team can’t address every issue as most are not a priority. But there is a massive effort right now to evaluate PrS submitted by contributors.

**Comment by nero2009** [comment]
I have been struggling to understand why my chat was lagging when tested on a Samsung S20 ultra with 12gb of RAM while it works prefectly on my xiaomi 9 with just 3GB of RAM. And the comments in this thread helped me fix the issue. 
Thanks everyone🎉

**Comment by neeraj-nh** [comment]
I am facing the same issue and have applied the solution here : https://github.com/facebook/react-native/issues/30034#issuecomment-780547496 .  Hoping it does not break anything else. I'll post here if it does 🤞 .

**Comment by sfuqua** [comment]
Encountered this issue in a chat scenario (where inverting the list is desirable) and was pulling my hair out why adding text elements to the list was causing scroll perf to tank.

Can confirm that uninverting the list and applying the scaleY workaround fixes scroll perf, but breaks the pull-to-refresh functionality :( may need to do something custom there.

Does anyone have any idea at an Android level why the perf implodes with the default implementation?

**Comment by JamesMahy** [comment]
> Encountered this issue in a chat scenario (where inverting the list is desirable) and was pulling my hair out why adding text elements to the list was causing scroll perf to tank.
> 
> Can confirm that uninverting the list and applying the scaleY workaround fixes scroll perf, but breaks the pull-to-refresh functionality :( may need to do something custom there.
> 
> Does anyone have any idea at an Android level why the perf implodes with the default implementation?
> 
> React Native 0.66 on a Pixel 6 / Android 12.

It was only something I saw on phones using One UI 3 so surprised it's now showing up on Android 12 too

**Comment by enahum** [comment]
> Encountered this issue in a chat scenario (where inverting the list is desirable) and was pulling my hair out why adding text elements to the list was causing scroll perf to tank.
> 
> 
> 
> Can confirm that uninverting the list and applying the scaleY workaround fixes scroll perf, but breaks the pull-to-refresh functionality :( may need to do something custom there.
> 
> 
> 
> Does anyone have any idea at an Android level why the perf implodes with the default implementation?
> 
> 
> 
> React Native 0.66 on a Pixel 6 / Android 12.
> 
> 
> 
> edit: between this issue and https://github.com/facebook/react-native/issues/30373 I feel that inverted FlatLists are just completely unusable on Android out-of-the-box - my team will likely need to invest in a custom list control for our chat canvas.

@sfuqua you may want to take a peek at https://github.com/mattermost/mattermost-mobile/blob/master/app/components/post_list/post_list.tsx and https://github.com/mattermost/mattermost-mobile/blob/master/app/components/post_list/post_list_refresh_control.tsx for an inverted list with working refresh control using scaleY: -1

**Comment by sfuqua** [comment]
> > Encountered this issue in a chat scenario (where inverting the list is desirable) and was pulling my hair out why adding text elements to the list was causing scroll perf to tank.
> > Can confirm that uninverting the list and applying the scaleY workaround fixes scroll perf, but breaks the pull-to-refresh functionality :( may need to do something custom there.
> > Does anyone have any idea at an Android level why the perf implodes with the default implementation?
> > React Native 0.66 on a Pixel 6 / Android 12.
> > edit: between this issue and #30373 I feel that inverted FlatLists are just completely unusable on Android out-of-the-box - my team will likely need to invest in a custom list control for our chat canvas.
> 
> @sfuqua you may want to take a peek at https://github.com/mattermost/mattermost-mobile/blob/master/app/components/post_list/post_list.tsx and https://github.com/mattermost/mattermost-mobile/blob/master/app/components/post_list/post_list_refresh_control.tsx for an inverted list with working refresh control using scaleY: -1

Cheers, I'll give that a try - I thought I'd tried that yesterday but may have been applying the wrong style to the RefreshControl.

**Comment by gorip1** [comment]
## Bottom refreshControl

For those who want to **keep the refresh control at the bottom**, you can wrap your flatlist in a view with scaleY:-1 instead of putting the style into the flat list, and invert the items. 

``` javascript
return (
<View style={{scaleY: -1}}>
  <FlatList
    // ...
    renderItem={({item}) => {
      return (
        <View style={{scaleY: -1}}>
          {/* cell content */}
        </View
      );
    }}
  />
</View>
);
```

**Comment by adamgajzlerowicz** [comment]
For me worked the following:
```
return (
<View style={{transform: [{ scaleY: -1 }]}}>
  <FlatList
    // ...
    renderItem={({item}) => {
      return (
        <View style={{transform: [{ scaleY: -1 }]}}>
          {/* cell content */}
        </View
      );
    }}
  />
</View>
);
```

**Comment by adamgajzlerowicz** [comment]
The `scaleY` style does not work on react 0.70.0

https://github.com/facebook/react-native/issues/34607

**Comment by adamgajzlerowicz** [comment]
But transform: [{ rotate: '180deg' }], does the trick!

**Comment by retrixe** [comment]
Worth noting this workaround isn't perfect, since the scroll indicator moves to the other side of the list (from the right to the left). Currently, I disabled the indicator in my application, but this is really not ideal.

The underlying problem here is that `scaleY` in `transform` is considerably slower than `scaleY` outside the `transform` property on Android for some reason, since `scaleY` is deprecated and not a web property anyway.

**Comment by retrixe** [comment]
For now, I have written a quick and dirty workaround to restore `scaleY` like so in my application's `index.js` file:
```js
// Add scaleY back to work around its removal in React Native 0.70.
import ViewReactNativeStyleAttributes from 'react-native/Libraries/Components/View/ReactNativeStyleAttributes'
ViewReactNativeStyleAttributes.scaleY = true
```
This allows you to continue using `scaleY` and avoid the aforementioned issue with the scroll bar being moved to the other side of the `FlatList` when using `rotate` instead of `scaleY`. This works fine in React Native 0.70.x, but might break in the future if support for the property is removed from the native Android `View` code.

**Comment by hotaryuzaki** [comment]
any solution with RN 0.70??
because scaleY StyleSheet is already remove from the package.

**Comment by holmesworcester** [comment]
Just noting here that we observed that in Android 13 the impact of this is much, much worse than in previous versions of Android, at least in our case (a long list of messages in a chat interface.) 

In Android 11 it was not so noticeable and in Android 13 it made the app completely unusable.

**Comment by fabriziobertoglio1987** [comment]
**Transform** 
The implementation of [transform][1] includes:
- Fabric C++ logic included in the [Transform.cpp][2] class (for ex. the implementation of [Translate][3])
- iOS implementation - [Transform.m][6]
- Android implementation - [BaseViewManager#setTransformProperty][5], [TransformHelper#processTransform][4] and for the translationX property the AOSP method [View#setTranslationX][7]

https://developer.android.com/reference/android/view/View#setScaleX(float)



[1]: https://reactnative.dev/docs/0.68/transforms#transform "reactnative transform documentation"
[2]: https://github.com/facebook/react-native/blob/873ff0c13b33d03af3060ecfc8fc9d31e0c5f589/ReactCommon/react/renderer/graphics/Transform.cpp "fabric c++ transform logic"
[3]: https://github.com/facebook/react-native/blob/873ff0c13b33d03af3060ecfc8fc9d31e0c5f589/ReactCommon/react/renderer/graphics/Transform.cpp#L58 "translate implementation in fabric c++"
[4]: https://github.com/facebook/react-native/blob/873ff0c13b33d03af3060ecfc8fc9d31e0c5f589/ReactAndroid/src/main/java/com/facebook/react/uimanager/TransformHelper.java#L94 "TransformHelper#processTransform is responsible for applying the tranlateX and translateY effects"
[5]: https://github.com/facebook/react-native/blob/873ff0c13b33d03af3060ecfc8fc9d31e0c5f589/ReactAndroid/src/main/java/com/facebook/react/uimanager/BaseViewManager.java#L447 "BaseViewManager#setTransformProperty"
[6]: https://github.com/facebook/react-native/blob/873ff0c13b33d03af3060ecfc8fc9d31e0c5f589/React/Views/RCTConvert+Transform.m#L14
[7]: https://github.com/aosp-mirror/platform_frameworks_base/blob/19e53cfdc8a5c6ef45c0adf2dd239576ddce5822/core/java/android/view/View.java#L17922 "AOSP method View#setTranslationX"

**Comment by fabriziobertoglio1987** [comment]
```
[
 Sx 0  0  0
 0  Sy 0  0
 0  0  Sz 0
 0  0  0  1
]
```
>A negative value (scaleY: -1) results in a [point reflection](https://en.wikipedia.org/wiki/Point_reflection) in that dimension. The value 1 has no effect.
>In [geometry](https://en.wikipedia.org/wiki/Geometry), a point reflection or inversion in a point (or inversion through a point, or central inversion) is a type of [isometry](https://en.wikipedia.org/wiki/Isometry) of [Euclidean space](https://en.wikipedia.org/wiki/Euclidean_space). 

<image src="https://upload.wikimedia.org/wikipedia/commons/5/57/Cyclic_symmetry_2.svg" width="200" />

Original: [Tomruen](https://commons.wikimedia.org/wiki/User:Tomruen) Vector: [KCVelaga](https://commons.wikimedia.org/wiki/User:KCVelaga)

Related [Transformation Matrix on Wikipedia](https://en.wikipedia.org/wiki/Transformation_matrix), [Scale Transformation](https://www.informit.com/articles/article.aspx?p=2854376&seqNum=8), [scale() on mdn docs](https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/scale) https://github.com/facebook/react-native/issues/30373#issuecomment-1173649280

**Comment by GuoXiaoyang** [comment]
Related issue: https://github.com/facebook/react-native/issues/35350

**Comment by Ashoat** [comment]
In case it's helpful to anybody, here's the latest that's necessary for an inverted list to play nice on Android 13 with React Native 0.70: https://phab.comm.dev/D6193

**Comment by swaroopbutala** [comment]
> In case it's helpful to anybody, here's the latest that's necessary for an inverted list to play nice on Android 13 with React Native 0.70: https://phab.comm.dev/D6193

Thank you! I have been struggling with this for the last few weeks! Thanks for putting in one place! It works great!

**Comment by Strate** [comment]
Also this issue could be fixed by also adding `scaleX: -1` in addition to `scaleY: -1` for container and every view.

https://github.com/facebook/react-native/issues/35350#issuecomment-1317465434

**Comment by zeallat** [comment]
Patch for RN 0.67.5 is here. Thanks for @pqkluan https://github.com/facebook/react-native/issues/30034#issuecomment-806396274

### react-native+0.67.5.patch
```diff
diff --git a/node_modules/react-native/Libraries/Lists/VirtualizedList.js b/node_modules/react-native/Libraries/Lists/VirtualizedList.js
index 2648cc3..a7e9f18 100644
--- a/node_modules/react-native/Libraries/Lists/VirtualizedList.js
+++ b/node_modules/react-native/Libraries/Lists/VirtualizedList.js
@@ -9,6 +9,7 @@
  */
 
 const Batchinator = require('../Interaction/Batchinator');
+const Platform = require('../Utilities/Platform');
 const FillRateHelper = require('./FillRateHelper');
 const ReactNative = require('../Renderer/shims/ReactNative');
 const RefreshControl = require('../Components/RefreshControl/RefreshControl');
@@ -2118,9 +2119,14 @@ function describeNestedLists(childList: {
 }
 
 const styles = StyleSheet.create({
-  verticallyInverted: {
-    transform: [{scaleY: -1}],
-  },
+  verticallyInverted:
+    Platform.OS === 'android'
+      ? {
+        scaleY: -1,
+      }
+      : {
+        transform: [{scaleY: -1}]
+      },
   horizontallyInverted: {
     transform: [{scaleX: -1}],
   },

```

**Comment by claudiozam** [comment]
What about RN 0.71? Same issue?

**Comment by MayheMatan** [comment]
any workarounds for react-native 0.71?

**Comment by claudiozam** [comment]
@MayheMatan try this

**index.js (in your app)**

```
import ViewReactNativeStyleAttributes from 'react-native/Libraries/Components/View/ReactNativeStyleAttributes';
ViewReactNativeStyleAttributes.scaleY = true;
```

**Then patch RN 0.71 (react-native+0.71.1.patch)**

```
diff --git a/node_modules/react-native/Libraries/Lists/VirtualizedList.js b/node_modules/react-native/Libraries/Lists/VirtualizedList.js
index 2629142..4e60aad 100644
--- a/node_modules/react-native/Libraries/Lists/VirtualizedList.js
+++ b/node_modules/react-native/Libraries/Lists/VirtualizedList.js
@@ -29,6 +29,7 @@ import {findNodeHandle} from '../ReactNative/RendererProxy';
 import flattenStyle from '../StyleSheet/flattenStyle';
 import StyleSheet from '../StyleSheet/StyleSheet';
 import clamp from '../Utilities/clamp';
+import Platform from '../Utilities/Platform';
 import infoLog from '../Utilities/infoLog';
 import {CellRenderMask} from './CellRenderMask';
 import ChildListCollection from './ChildListCollection';
@@ -1840,9 +1841,7 @@ export default class VirtualizedList extends StateSafePureComponent<
 }
 
 const styles = StyleSheet.create({
-  verticallyInverted: {
-    transform: [{scaleY: -1}],
-  },
+  verticallyInverted: Platform.OS === 'android' ? { scaleY: -1 } : { transform: [{scaleY: -1}] },
   horizontallyInverted: {
     transform: [{scaleX: -1}],
   },
```

**Comment by devoren** [comment]
Same problem for 0.71.2

**Comment by devoren** [comment]
@claudiozam it helps but only with a flat list (+JS drops to 30fps) but if using flashlist then UI and JS drops to 30fps :(

**Comment by claudiozam** [comment]
@devoren Flatlist is faster than Flashlist inverted.....

![image](https://user-images.githubusercontent.com/2105644/216331107-38f81d02-1760-47d6-8a09-d88f2046a379.png)

**Comment by devoren** [comment]
@claudiozam idk what do you mean?

**Comment by KoalaSat** [comment]
> For now, I have written a quick and dirty workaround to restore `scaleY` like so in my application's `index.js` file:
> 
> ```js
> // Add scaleY back to work around its removal in React Native 0.70.
> import ViewReactNativeStyleAttributes from 'react-native/Libraries/Components/View/ReactNativeStyleAttributes'
> ViewReactNativeStyleAttributes.scaleY = true
> ```
> 
> This allows you to continue using `scaleY` and avoid the aforementioned issue with the scroll bar being moved to the other side of the `FlatList` when using `rotate` instead of `scaleY`. This works fine in React Native 0.70.x, but might break in the future if support for the property is removed from the native Android `View` code.

This saved my day

**Comment by techyon7** [comment]
> solution

For me, it does solve the fps drop issue but the scrolling becomes quite janky. If I try to scroll fast, it would either stop abruptly, go in the opposite direction or scroll very unevenly. Is anyone else facing a similar issue?

**Comment by techyon7** [comment]
> I also encounter this issue with my chat screen. It's really hard to find to root cause.
> 
> * Only happen on Android S10e device.
> * With inverted FlatList.
> * FlatList Item contains a Text with any given length.
> 
> It's only take 10 items in the list for the performance to be hit really bad.
> 
> I'm using @cookiespam bandaid + `patch-package` to resolve the issue for now now.
> 
> Here is the the content of the patch:
> 
> `react-native+0.63.3.patch`
> 
> ```diff
> diff --git a/node_modules/react-native/Libraries/Lists/VirtualizedList.js b/node_modules/react-native/Libraries/Lists/VirtualizedList.js
> index 9ec105f..a8f1d8c 100644
> --- a/node_modules/react-native/Libraries/Lists/VirtualizedList.js
> +++ b/node_modules/react-native/Libraries/Lists/VirtualizedList.js
> @@ -11,6 +11,7 @@
>  'use strict';
>  
>  const Batchinator = require('../Interaction/Batchinator');
> +const Platform = require('../Utilities/Platform');
>  const FillRateHelper = require('./FillRateHelper');
>  const PropTypes = require('prop-types');
>  const React = require('react');
> @@ -2185,9 +2186,14 @@ function describeNestedLists(childList: {
>  }
>  
>  const styles = StyleSheet.create({
> -  verticallyInverted: {
> -    transform: [{scaleY: -1}],
> -  },
> +  verticallyInverted:
> +    Platform.OS === 'android'
> +      ? {
> +          scaleY: -1,
> +        }
> +      : {
> +          transform: [{scaleY: -1}]
> +        },
>    horizontallyInverted: {
>      transform: [{scaleX: -1}],
>    },
> ```

For me, it does solve the fps drop issue but the scrolling becomes quite janky. If I try to scroll fast, it would either stop abruptly, go in the opposite direction or scroll very unevenly. Is anyone else facing a similar issue?

**Comment by rkyle35242** [comment]
Throwing a transform style on the flatlist and children of `scaleY: -1` also had performance issues for us. However, rotating the list and children `180deg` didn't seem to have the same problem and kept the "inverted" functionality.

**Comment by tj-mc** [comment]
Still facing this in 0.71. This is an issue for all chat-like applications

**Comment by Sondago** [comment]
> Still facing this in 0.71. This is an issue for all chat-like applications

+1

**Comment by karthik-balasubramanyam15** [comment]
+1
I even tried to use Animated Section List instead of chat list and still facing the same issue with inverted.

**Comment by thanksyouall** [comment]
I have a problem that any inverted list on Android slows down when scrolling. It's not freezing, it's the touch processing. If I try to scroll quickly with my finger, I can't do it. I don't have this problem on a non-flipped list, though. The same problem in ScrollView.

**Comment by fukemy** [comment]
still problem in react native 0.71.8, so sad

**Comment by claudiozam** [comment]
In React Native 0.72.3 the issue persist. Now the VirtualizedList is in another location @react-native/virtualized-lists/List

**Comment by anatooly** [comment]
npx patch-package @react-native/virtualized-lists

patches/@react-native+virtualized-lists+0.72.6.patch
```
diff --git a/node_modules/@react-native/virtualized-lists/Lists/VirtualizedList.js b/node_modules/@react-native/virtualized-lists/Lists/VirtualizedList.js
index 8a23e81..b2be91a 100644
--- a/node_modules/@react-native/virtualized-lists/Lists/VirtualizedList.js
+++ b/node_modules/@react-native/virtualized-lists/Lists/VirtualizedList.js
@@ -23,6 +23,7 @@ import type {
   RenderItemType,
   Separators,
 } from './VirtualizedListProps';
+import Platform from '../Utilities/Platform';
 
 import {
   RefreshControl,
@@ -2056,9 +2057,7 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
 }
 
 const styles = StyleSheet.create({
-  verticallyInverted: {
-    transform: [{scaleY: -1}],
-  },
+  verticallyInverted: Platform.OS === 'android' ? { scaleY: -1 } : { transform: [{scaleY: -1}] },
   horizontallyInverted: {
     transform: [{scaleX: -1}],
   },
```

**Comment by Andrija00** [comment]
In react native versions 0.71+ to stop performance drop on Android turn off inverted FlatList and add this : style={{transform: [{rotate: '180deg'}]}}, also in add this style in renderItem that is in flat list., also add this if you want to add scroll bar on the right side you can use showsVerticalScrollIndicator={false}. This fixed our chat app where we  had lag on many android device when we have 10k char messages.

**Comment by fukemy** [comment]
@Andrija00 dont do this, instead of styles flatlist and component just styles flatlist 1 time by:
```
diff --git a/node_modules/@react-native/virtualized-lists/Lists/VirtualizedList.js b/node_modules/@react-native/virtualized-lists/Lists/VirtualizedList.js
index 8a23e81..ca17f79 100644
--- a/node_modules/@react-native/virtualized-lists/Lists/VirtualizedList.js
+++ b/node_modules/@react-native/virtualized-lists/Lists/VirtualizedList.js
@@ -30,6 +30,7 @@ import {
   View,
   StyleSheet,
   findNodeHandle,
+  Platform,
 } from 'react-native';
 import Batchinator from '../Interaction/Batchinator';
 import clamp from '../Utilities/clamp';
@@ -2056,9 +2057,7 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
 }
 
 const styles = StyleSheet.create({
-  verticallyInverted: {
-    transform: [{scaleY: -1}],
-  },
+  verticallyInverted: Platform.OS === 'android' ? { scaleY: -1 } : { transform: [{scaleY: -1}] },
   horizontallyInverted: {
     transform: [{scaleX: -1}],
   },
```

**Comment by Andrija00** [comment]
> @Andrija00 dont do this, instead of styles flatlist and component just styles flatlist 1 time by:
> 
> ```
> diff --git a/node_modules/@react-native/virtualized-lists/Lists/VirtualizedList.js b/node_modules/@react-native/virtualized-lists/Lists/VirtualizedList.js
> index 8a23e81..ca17f79 100644
> --- a/node_modules/@react-native/virtualized-lists/Lists/VirtualizedList.js
> +++ b/node_modules/@react-native/virtualized-lists/Lists/VirtualizedList.js
> @@ -30,6 +30,7 @@ import {
>    View,
>    StyleSheet,
>    findNodeHandle,
> +  Platform,
>  } from 'react-native';
>  import Batchinator from '../Interaction/Batchinator';
>  import clamp from '../Utilities/clamp';
> @@ -2056,9 +2057,7 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
>  }
>  
>  const styles = StyleSheet.create({
> -  verticallyInverted: {
> -    transform: [{scaleY: -1}],
> -  },
> +  verticallyInverted: Platform.OS === 'android' ? { scaleY: -1 } : { transform: [{scaleY: -1}] },
>    horizontallyInverted: {
>      transform: [{scaleX: -1}],
>    },
> ```

**Comment by Andrija00** [comment]
I see but I am not using Expo or Virtualized List. I am using basic flatList from React Native and rotate will fix the issue. Thanks for help!

**Comment by hannojg** [comment]
`VirtualizedList` is the base component that `FlatList` uses to implement the list.

**Comment by anatooly** [comment]
@Andrija00 are you check using FlashList for you chat (10k char messages)?

**Comment by chelovekdrakon** [comment]
@hannojg do you have any idea when it will go out?
or should I use `main` branch?
Looks like `ViewReactNativeStyleAttributes.scaleY = true;` does not work anymore on 0.72 with new arch - just updated ;/

**Comment by Andrija00** [comment]
@anatooly We had issues using FlashList in the past for your ChatScreen component so we stick with FlatList

**Comment by Nasseratic** [comment]
@hannojg great work!! 💪 😍

**Comment by hannojg** [comment]
The fix is part of [RN 0.72.4](https://github.com/facebook/react-native/releases/tag/v0.72.4), you can upgrade and test 😊

**Comment by retrixe** [comment]
Perhaps this issue should be closed?

**Comment by kelset** [comment]
if anyone can confirm that it's fixed I'd be happy to close

**Comment by daidonpah** [comment]
Unfortunately for me updating did not resolve the issue. On Samsung S23 Ultra Android 13, when removing the `inverted` prop, FlatList behaves perfectly fine as it should, running at smooth 120fps. Adding the `inverted` prop sporadically brings the fps down to 0.5 and a lot of frame drops. After updating to RN 0.72.4 it seemed liked the issue "improved" at first, but after 20 seconds of scrolling, same performance drop again.

**Comment by daidonpah** [comment]
Nevermind my previous comment, I had tried to play around with FlashList instead and forgotten to change it back to FlatList. By using FlatList the issue does indeed seem to be fixed, _although_ now the scroll indicator is on the **left**.

**Comment by hannojg** [comment]
@daidonpah Are you suer your flatlist doesn't have any custom styles? Because that may cause the wrong position of the scrollbar. 

If you don't, are you able to reproduce in a clean new app? If so, I think I'd be good to open a new issue 😊

**Comment by mirkos93** [comment]
@daidonpah Have you solved the issue using FlashList?

**Comment by foloinfo** [comment]
Looks like the performance issue is fixed with [v0.72.4](https://github.com/facebook/react-native/releases/tag/v0.72.4) with [this line](https://github.com/facebook/react-native/blob/3dd816c6b7bcd9cc4c21199f0b645755fb97f50f/packages/virtualized-lists/Lists/VirtualizedList.js#L1980) using `{transform: [{scale: -1}]}` not `scaleY`.
Thank you for the fix @hannojg

I'm using Expo 48 and it uses `"react-native": "0.71.14"` so I had to patch myself.
This will hide the scrollbar for Android, but better than huge performance drop while I wait to update RN.

```javascript
const styles = StyleSheet.create({
  inverted: {
    ...Platform.select({
      android: {
        transform: [{ scale: -1 }],
      },
    })
  }
});

...

return (
  <FlatList
    // inverted={true} on Android 33+ will cause performance issue
    inverted={Platform.OS === 'ios'}
    // apply the inverted style on android
    style={styles.inverted}
    // hide the scroll indicator on android because it shows on the left
    showsVerticalScrollIndicator={Platform.OS === 'ios'}
    listFooterComponent={
      <View style={styles.inverted}>
        <Header/>
      </View>
    }
    renderItem={({ item }) => {
      return (
        <View style={styles.inverted}>
          {renderItem(item)}
        </View>
      )
    }}
  />
);
```

**Comment by retrixe** [comment]
@foloinfo on earlier versions this workaround works without compromising the scrollbar:
https://github.com/facebook/react-native/issues/30034#issuecomment-780547496
https://github.com/facebook/react-native/issues/30034#issuecomment-1277360480

**Comment by JJSLIoT** [comment]
Not stale!

**Comment by elias96** [comment]
Keep this open amigo

**Comment by retrixe** [comment]
This issue has been fixed in recent versions of React Native unless you are still experiencing it, in which case it may be more appropriate to open a new issue

> The fix is part of [RN 0.72.4](https://github.com/facebook/react-native/releases/tag/v0.72.4), you can upgrade and test 😊
